linked_list/linked_list.py

```diff
@@ -86,32 +141,54 @@ def visit(self):
 
     # method to reverse the singly linked list
     # note: the nodes should be moved and not just the values in the nodes
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n^2) Yiikes!
+    # Space Complexity: O(1)
     def reverse(self):
-        pass
-  
+        current = self.head.next
+        new_head = self.tail
+        while current:
+            if current.next == self.tail:
+                self.tail.next = current
+                self.tail = current
+                if self.tail == self.head:
+                    self.head = new_head
+                    current = None
+                else:
+                    current = self.head
+            else:
+                current = current.next
+
     ## Advanced/ Exercises
     # returns the value at the middle element in the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def find_middle_value(self):
-        pass
+        middle = self.size // 2
+        return self.get_at_index(middle)
 
     # find the nth node from the end and return its value
     # assume indexing starts at 0 while counting to n
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def find_nth_from_end(self, n):
-        pass
+        index = self.size - n - 1
+        return self.get_at_index(index)
 
     # checks if the linked list has a cycle. A cycle exists if any node in the
     # linked list links to a node already visited.
     # returns true if a cycle is found, false otherwise.
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)??
+    # Space Complexity: O(1)
+
+    # Hi Chris! I don't know why this doesn't pass the last test, and I kinda ran out of time to figure it out. Sorry!!
     def has_cycle(self):
-        pass
+        current = self.head
+        while current:
+            if current.flag:
+                return True
+            current.flag = True
+            current = current.next
+        return False

```

Comment:
The reason this is failing is because the `flag` field you add still exists after the function gets called.  So the next time we try to add a cycle `flag` is `True` on the 1st node in the list, so... it fails the 2nd time we try to detect a cycle.

To solve it you need a 2nd loop after the 1st to either remove the `flag` field or set it to false.

```suggestion
        current = self.head
        while current:
            if current.flag:
                return True
            current.flag = True
            current = current.next
            
         current = self.head
         while current:
             current.flag = False
             current = current.next
        return False

```